spritecontrol.cpp

```diff
@@ -391,6 +392,247 @@ std::map<PatternName, GlobalPlayer::MoveFunction> GlobalPlayer::move_functions {
+				double relative_angle = get_angle(diff);
+				relative_angle = wrap_angle(relative_angle - current_angle);	// 180 deg: straight ahead; 0 deg: straight behind (assuming i'm not bad at math)
+
+				if (abs(relative_angle) >= blind_angle[current_sprite->id()]) {
+					sprites_seen++;
+					double other_angle = get_angle(velocity[other_sprite->id()]);
+					average_angle += other_angle;
+					average_pos += other_final;
+				}
+
+				if (dist < desired_separation[current_sprite->id()]) {
+					diff /= dist * dist / desired_separation[current_sprite->id()];
+					separation_velocity_changes[current_sprite] = diff;
+				}
+			}
+
+			if (sprites_seen > 1) {
+				average_angle /= sprites_seen;
+				average_pos /= (double) sprites_seen;
+
+				// i sure hope the angles i'm calculating line up with the angles used for movement
+				alignment_velocity_changes[current_sprite] = Point(cos(average_angle), sin(average_angle));
+
+				Point relative_average_pos = average_pos - Point(current_sprite->final<X>(), current_sprite->final<Y>());
+				cohesion_velocity_changes[current_sprite] = normalize_vector(relative_average_pos);
+			}
+
+			Point edge_push = Point(0.0, 0.0);
+
+			if (-1.0 + EDGE_PUSH_X_RADIUS > get<X>(current_final)) {
+				get<X>(edge_push) += ((-1.0 + EDGE_PUSH_X_RADIUS) - get<X>(current_final)) / EDGE_PUSH_X_RADIUS;
+			} else if (1.0 - EDGE_PUSH_X_RADIUS < get<X>(current_final)) {
+				get<X>(edge_push) += ((1.0 - EDGE_PUSH_X_RADIUS) - get<X>(current_final)) / EDGE_PUSH_X_RADIUS;
+			}
+
+			if (-1.0 + EDGE_PUSH_X_RADIUS > get<Y>(current_final)) {
+				get<Y>(edge_push) += ((-1.0 + EDGE_PUSH_X_RADIUS) - get<Y>(current_final)) / EDGE_PUSH_X_RADIUS;
+			} else if (1.0 - EDGE_PUSH_X_RADIUS < get<Y>(current_final)) {
+				get<Y>(edge_push) += ((1.0 - EDGE_PUSH_X_RADIUS) - get<Y>(current_final)) / EDGE_PUSH_X_RADIUS;
+			}
+
+			if (edge_push != Point(0.0, 0.0)) {
+				get<X>(edge_push) = pow(abs(get<X>(edge_push)), 1.0/2.5) * (get<X>(edge_push) < 0.0 ? -1.0 : 1.0);
+				get<Y>(edge_push) = pow(abs(get<Y>(edge_push)), 1.0/2.5) * (get<Y>(edge_push) < 0.0 ? -1.0 : 1.0);
+				edge_push_velocity_changes[current_sprite] = edge_push;
+			}
+		}
+
+		for (auto &sprite : separation_velocity_changes) {
+			double scale = get_vector_magnitude(sprite.second);
+			scale = min(desired_velocity[sprite.first->id()] / scale, 1.0) * SEPARATION_FORCE_MULT;
+			velocity[sprite.first->id()] += sprite.second * scale;
+		}
+
+		for (auto &sprite : alignment_velocity_changes) {
+			Point &sprite_velocity = velocity[sprite.first->id()];
+			double magnitude = get_vector_magnitude(sprite_velocity);
+			sprite_velocity /= magnitude;
+
+			Point diff = sprite.second - sprite_velocity;
+			diff *= ALIGNMENT_FORCE_MULT;
+
+			sprite_velocity = normalize_vector(sprite_velocity + diff) * magnitude;
+			// there's probably a better way to do this but i can't be bothered figuring it out now
+		}
+
+		for (auto &sprite : cohesion_velocity_changes) {
```

Comment:
More on structured binding: you can write the for loop like this:
```cpp
for (auto &[sprite, velocity_change] : alignment_velocity_changes) {
```
And then what used to be `sprite.first` becomes `sprite` and what used to be `sprite.second` becomes `velocity_change`. Basically, it gives you nicer variable names for free, which we should definitely use _if_ these for loops continue to exist.